cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,220 @@
+                                          {column_order}, {null_precedence}, mr,
+                                          stream)
+            : detail::sorted_order(table_view{{input_col}}, {column_order},
+                                   {null_precedence}, mr, stream);
+    column_view sorted_order_view = sorted_order->view();
+
+    // na_option=keep assign NA to NA values
+    if (_include_nulls == include_nulls::NO)
+      rank_columns.push_back(make_numeric_column(
+          data_type(FLOAT64), size, copy_bitmask(input_col, stream, mr),
+          input_col.null_count(), stream, mr));
+    else
+      rank_columns.push_back(make_numeric_column(
+          data_type(FLOAT64), size, mask_state::UNALLOCATED, stream, mr));
+
+    auto rank_mutable_view = rank_columns.back()->mutable_view();
+    auto rank_data = rank_mutable_view.data<double>();
+    auto device_table =
+        table_device_view::create(table_view{{input_col}}, stream);
+
+    // All equal values have same rank and rank always increases by 1 between
+    // groups
+    rmm::device_vector<size_type> dense_rank_sorted; // as key for min, max,
+                                                     // average
+    if (method != rank_method::FIRST) {
+      dense_rank_sorted = rmm::device_vector<size_type>(size);
+      if (input_col.has_nulls()) {
+        auto conv = unique_comparator<true, size_type>(
+            *device_table, sorted_order_view.data<size_type>());
+        auto it = thrust::make_transform_iterator(
+            thrust::make_counting_iterator<size_type>(0), conv);
+        thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream), it,
+                               it + size,
+                               dense_rank_sorted.data().get());
+      } else {
+        auto conv = unique_comparator<false, size_type>(
+            *device_table, sorted_order_view.data<size_type>());
+        auto it = thrust::make_transform_iterator(
+            thrust::make_counting_iterator<size_type>(0), conv);
+        thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream), it,
+                               it + size,
+                               dense_rank_sorted.data().get());
+      }
+    }
+
+    switch (method) {
+    case rank_method::FIRST:
+      // stable sort order ranking (no ties)
+      thrust::scatter(
+          rmm::exec_policy(stream)->on(stream),
+          thrust::make_counting_iterator<double>(1),
+          thrust::make_counting_iterator<double>(size + 1),
+          sorted_order_view.begin<size_type>(), rank_data);
+      break;
+    case rank_method::DENSE:
+      // All equal values have same rank and rank always increases by 1 between
+      // groups
+      thrust::scatter(rmm::exec_policy(stream)->on(stream),
+                      dense_rank_sorted.begin(), dense_rank_sorted.end(),
+                      sorted_order_view.begin<size_type>(), rank_data);
+      break;
+    case rank_method::MIN: {
```

Comment:
The MIN and MAX methods are nearly identical save for using `thrust::minimum` vs `thrust::maxium`. Consider refactoring into a common function.

Reply:
refactored to `tie_break_ranks` (min, max) and `tie_break_ranks_transform` (average).